webservices/server-integration/src/main/java/org/jboss/as/webservices/deployers/WSIntegrationProcessorJAXWS_JMS.java

```diff
@@ -82,48 +83,56 @@ public void deploy(final DeploymentPhaseContext phaseContext) throws DeploymentU
             //support for contract-first development only: pick-up @WebService annotations referencing a provided wsdl contract only
             if (wsdlLocation != null && port != null && service != null) {
                 String key = wsdlLocation.asString();
-                List<AnnotationInstance> annotations = map.get(key);
-                if (annotations == null) {
-                    annotations = new LinkedList<AnnotationInstance>();
-                    map.put(key, annotations);
-                }
+                List<AnnotationInstance> annotations = map.computeIfAbsent(key, k -> new LinkedList<>());
                 annotations.add(webServiceAnnotation);
             }
         }
 
-        //extract SOAP-over-JMS 1.0 bindings
-        List<JMSEndpointMetaData> list = new LinkedList<JMSEndpointMetaData>();
-        if (!map.isEmpty()) {
-
-            for (String wsdlLocation : map.keySet()) {
-                try {
-                    final ResourceRoot resourceRoot = getWsdlResourceRoot(unit, wsdlLocation);
-                    if (resourceRoot == null) continue;
-                    final VirtualFile wsdlLocationFile = resourceRoot.getRoot().getChild(wsdlLocation);
-                    final URL url = wsdlLocationFile.toURL();
-                    SOAPAddressWSDLParser parser = new SOAPAddressWSDLParser(url);
-                    for (AnnotationInstance ai : map.get(wsdlLocation)) {
-                        String port = ai.value(PORT_NAME).asString();
-                        String service = ai.value(SERVICE_NAME).asString();
-                        AnnotationValue targetNS = ai.value(TARGET_NAMESPACE);
-                        String tns = targetNS != null ? targetNS.asString() : null;
-                        QName serviceName = new QName(tns, service);
-                        QName portName = new QName(tns, port);
-                        String soapAddress = parser.filterSoapAddress(serviceName, portName, SOAPAddressWSDLParser.SOAP_OVER_JMS_NS);
-                        if (soapAddress != null) {
-                            ClassInfo webServiceClassInfo = (ClassInfo) ai.target();
-                            String beanClassName = webServiceClassInfo.name().toString();
-                            //service name ?
-                            list.add(new JMSEndpointMetaData(beanClassName, port, beanClassName, wsdlLocation, soapAddress));
-                        }
+        final List<JMSEndpointMetaData> list = extractSOAPOverJMSBindings(unit, map);
```

Comment:
Maybe it should have always been a read only list but at this point the list returned could now be read only OR modifiable.